### PR TITLE
add type coverage report tool

### PR DESCRIPTION
**Pull request by sokra**
**What kind of change does this PR introduce?**
tooling

**Did you add tests for your changes?**
N/A

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
nothing

### CODE REVIEW

File: tooling/type-coverage.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+					const typeText = typeChecker.typeToString(type);
+					let isExtenalFunction = false;
+					if (/^(<.*>)?\(/.test(typeText) && type.symbol) {
+						for (const decl of type.symbol.getDeclarations()) {
+							const sourceFile = decl.getSourceFile();
+							if (sourceFile && !isProjectFile(sourceFile.fileName))
+								isExtenalFunction = true;
+						}
+					}
+					/*console.log(
```
Comment (ooflorent) [inline tooling/type-coverage.js]: Could you remove this?

File: tooling/type-coverage.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+					const {
+						line: lineEnd,
+						character: characterEnd
+					} = ts.getLineAndCharacterOfPosition(sourceFile, node.getEnd());
+					const typeText = typeChecker.typeToString(type);
+					let isExtenalFunction = false;
+					if (/^(<.*>)?\(/.test(typeText) && type.symbol) {
+						for (const decl of type.symbol.getDeclarations()) {
+							const sourceFile = decl.getSourceFile();
+							if (sourceFile && !isProjectFile(sourceFile.fileName))
```
Comment (ooflorent) [inline tooling/type-coverage.js]: You could break if the test succeeds

File: tooling/type-coverage.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+					const { line, character } = ts.getLineAndCharacterOfPosition(
+						sourceFile,
+						node.getStart()
+					);
+					const {
+						line: lineEnd,
+						character: characterEnd
+					} = ts.getLineAndCharacterOfPosition(sourceFile, node.getEnd());
+					const typeText = typeChecker.typeToString(type);
+					let isExtenalFunction = false;
```
Comment (ooflorent) [inline tooling/type-coverage.js]: Typo:

```diff
- isExtenalFunction
+ isExternalFunction

File: tooling/type-coverage.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+							if (sourceFile && !isProjectFile(sourceFile.fileName))
+								isExtenalFunction = true;
+						}
+					}
+					/*console.log(
+						`${file}:${line+1}:${character+1} ${node.getText()} ${typeText} ${isExtenalFunction}`
+					);*/
+					const isTyped =
+						isExtenalFunction ||
+						(!(type.flags & ts.TypeFlags.Any) && !/\bany\b/.test(typeText));
```
Comment (ooflorent) [inline tooling/type-coverage.js]: Should we also skip `TODO`?

Comment (sokra) [reply tooling/type-coverage.js]: TODO resolves to `any`. It's handled like `any`

**Pull request closed without merging**